packages/xod-project/src/messages.js

```diff
@@ -64,6 +64,78 @@ export default {
       conflictingSpecializations
     )}`,
   }),
+  SPECIALIZATION_PATCH_CANT_BE_ABSTRACT: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: "Specialization patches can't be abstract",
+    trace,
+  }),
+  SPECIALIZATION_PATCH_MUST_HAVE_SAME_ARITY_LEVEL: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: 'Specialization patches must have same arity level as abstract',
+    trace,
+  }),
+  SPECIALIZATION_PATCH_CANT_HAVE_GENERIC_PINS: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: "Specialization patches can't have generic pins",
+    trace,
+  }),
+  SPECIALIZATION_PATCH_MUST_HAVE_N_INPUTS: ({
+    abstractPatchPath,
+    desiredInputsNumber,
+    trace,
+  }) => ({
+    title: 'Invalid specialization patch',
+    note: `Specialization patch for ${abstractPatchPath} must have ${desiredInputsNumber} inputs`,
+    trace,
+  }),
+  SPECIALIZATION_PATCH_MUST_HAVE_N_OUTPUTS: ({
+    desiredOutputsNumber,
+    abstractPatchPath,
+    trace,
+  }) => ({
+    title: 'Invalid specialization patch',
+    note: `Specialization patch for ${abstractPatchPath} must have ${desiredOutputsNumber} outputs`,
+    trace,
+  }),
+  SPECIALIZATION_STATIC_PINS_DO_NOT_MATCH: ({ abstractPatchPath, trace }) => ({
+    title: 'Invalid specialization patch',
+    note: `Static pins do not match with ${abstractPatchPath}`,
+    trace,
+  }),
+  SPECIALIZATION_HAS_CONFLICTING_TYPES_FOR_GENERIC: ({
+    genericType,
+    typeNames,
+    trace,
+  }) => ({
+    title: 'Invalid specialization patch',
+    note: `Conflicting types for generic ${genericType}: ${typeNames}`,
+    trace,
+  }),
+  SPECIALIZATION_HAS_WRONG_NAME: ({
+    expectedSpecializationBaseName,
+    trace,
+  }) => ({
+    title: 'Invalid specialization patch',
+    note: `Patch should be named ${expectedSpecializationBaseName}`,
```

Comment:
Here and above, solutions are missing. We should provide them to be a friendlier environment.